source/extensions/filters/common/rbac/matchers.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/rbac/v2alpha/rbac.pb.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/connection.h"
+
+#include "common/http/header_utility.h"
+#include "common/network/cidr_range.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RBAC {
+
+class Matcher;
+typedef std::shared_ptr<const Matcher> MatcherConstSharedPtr;
+
+/**
+ *  Matchers describe the rules for matching either a permission action or principal.
+ */
+class Matcher {
+public:
+  virtual ~Matcher() {}
+
+  /**
+   * Returns whether or not the connection matches the rules of the matcher. This overload of
+   * `matches` should only be used for non-HTTP network calls (e.g., a binary TCP protocol).
+   *
+   * @param connection the downstream connection used to match against.
+   */
+  virtual bool matches(const Network::Connection& connection) const PURE;
+
+  /**
+   * Returns whether or not the HTTP request matches the rules of the matcher. This overload of
+   * `matches` should only be used for HTTP requests.
+   *
+   * @param connection the downstream connection used to match against.
+   * @param headers    the request headers used to match against.
+   */
+  virtual bool matches(const Network::Connection& connection,
+                       const Envoy::Http::HeaderMap& headers) const PURE;
+
+  /**
+   * Creates a shared instance of a matcher based off the rules defined in the Permission config
+   * proto message.
+   */
+  static MatcherConstSharedPtr create(const envoy::config::rbac::v2alpha::Permission&);
```

Comment:
nit: for function decls, spell out the parameter name (this will end up getting caught by clang-tidy on google import which we don't run in public CI). (Same elsewhere.)